backends/dpdk/dpdkHelpers.cpp

```diff
@@ -681,9 +682,52 @@ bool ConvertStatementToDpdk::preorder(const IR::MethodCallStatement *s) {
     return false;
 }
 
-// TODO(hanw): TBD
-bool ConvertStatementToDpdk::preorder(const IR::SwitchStatement *) {
-    BUG("Not implemented");
+// This function assumes EliminateSwitch midend pass is applied and only handles
+// action_run in switch expression.
+bool ConvertStatementToDpdk::preorder(const IR::SwitchStatement *s) {
+    auto tc = P4::TableApplySolver::isActionRun(s->expression, refmap, typemap);
+    BUG_CHECK(tc != nullptr, "%1%: unexpected switch statement expression",
+              s->expression);
+    auto size = s->cases.size();
+    std::vector <cstring> labels;
+    cstring label;
+    bool isdefaultPresent = false;
+    cstring default_label = "";
+    auto end_label = Util::printf_format("label_endswitch%d", next_label_id++);
+    // Emit jmp on action run statements and collect labels for each case statement
+    for (unsigned i = 0; i < size; i++) {
+        auto caseLabel = s->cases.at(i);
+        if (caseLabel->label->is<IR::DefaultExpression>()) {
+            label = Util::printf_format("label_default%d", next_label_id++);
+            add_instr(new IR::DpdkJmpLabelStatement(label));
+            default_label = label;
+            isdefaultPresent = true;
+        } else {
+            auto pe = caseLabel->label->to<IR::PathExpression>();
+            CHECK_NULL(pe);
+            label = Util::printf_format("label_action%d", next_label_id++);
+            add_instr(new IR::DpdkJmpOnActionRunStatement(label, pe->path->name.name));
+        }
+        labels.push_back(label);
+    }
+
+    if (!isdefaultPresent)
```

Comment:
don't we have a pass which ensures that this is always true (or that the cases are exhaustive?)

Reply:
There is a frontend pass for adding default case. However, later the midend pass EliminateSwitch converts switch statements with other expressions to switch case with action_run expression. These newly created switch statements do not have a default case, the default case is also converted to an action and that action is marked as the const default action for the switch table.
https://github.com/p4lang/p4c/blob/main/midend/eliminateSwitch.h#L73

Reply:
One solution is then to rerun this pass after the switch statements are eliminated...